### PR TITLE
fix(corednsManifest): add recommended labels

### PR DESCRIPTION
**Pull request by paologallinaharbur**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Kubernetes recommends [some labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/), but it does not apply them to coreDNS. It should have at least the app.kubernetes.io/name one

Following [coredns chart](https://github.com/coredns/helm/blob/0dbb02cedec3851e19031cde10f63901bbaf64bb/charts/coredns/templates/service.yaml#L16), but avoiding to add it to `deployment.spec.selector.matchLabels[]` since field is immutable, not sure if we should to consider upgrade path.

It would also make sense to add them to the kube-proxy DaemonSet, let me know what you think

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
hi, this is debatable but the `app.kubernetes.io/name` label is more about applications deployed by the user and not about system components like coredns. it has been discussed before and i was -1 on the addition of this label to the kubeadm deployed components inc. static pods and "core" addons.

**Comment by paologallinaharbur** [comment]
I understand your point.
In our use case, we leverage `app.kubernetes.io/name` label to discriminate between the different technology exposing Prometheus metrics auto-discovered in the cluster. 
Since metrics scraped can be colliding (think for example to all the go_, process_ ones) or simply not being correctly namespaced it is complex to build dashboards without relying on a common label such `app.kubernetes.io/name`.

**Comment by neolit123** [comment]
users can always apply these post 'kubeadm init'. one concern would be if these core components suddenly start appearing in metrics where they are not needed.

for kubeadm we always follow the minimal config rule, so if something is not needed by all users, perhaps its better to leave it opt in.

**Comment by neolit123** [comment]
of course if there are more opinions by the other maintainers lets discuss it. cc @SataQiu @pacoxu

**Comment by pacoxu** [comment]
I would let coredns maintainers decide it.
I see https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed is still not using app labels.

`app.kubernetes.io/name` is in https://kubernetes.io/docs/reference/labels-annotations-taints/ and https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/. I would suggest adding it to https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed  at first and syncing to kubeadm if this is best practice.  It seems to be already a helm best practice according to [coredns/helm](https://github.com/coredns/helm/).

**Comment by paologallinaharbur** [comment]
Thanks for pointing out the deployment repo, I opened a PR to add `app.kubernetes.io/name` and it was merged. 
Now both the helm chart and coredns manifest have such label, therefore I would add here also 😃

**Comment by paologallinaharbur** [comment]
(At least from my point of view) Since there are already non-standard labels such as `kubernetes.io/name: "CoreDNS"` or the ones related to the Prometheus server, it makes sense to add the standard ones.

For example, having an opt-in configuration would be problematic for our use case since leverage autodiscovery: users would get the metrics without a standard label to rely on.

**Comment by neolit123** [comment]
> For example, having an opt-in configuration would be problematic for our use case since leverage autodiscovery: users would get the metrics without a standard label to rely on.

the cluster admin could use kubectl to patch the label after kubeadm init.

no strong opinions on my end as long as we are not breaking user expectations i.e. kubeadm addons don't have the labels by default.

**Comment by paologallinaharbur** [comment]
Hello again, yes, the admin could add such label manually. 
However, out-of-the-box the metric would be already scraped if Prometheus is installed (due to the `prometheus.io/scrape: "true"` annotation) and not labeled with a standard label, but with not-standard ones:
 - `kubernetes.io/name: "CoreDNS"`
 -  `k8s-app: kube-dns`

@pacoxu, can we move this forward or the neolit123 concern is a blocker?

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (sftim) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go:66]: I would set `app.kubernetes.io/component` too.

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (sftim) [inline cmd/kubeadm/app/phases/addons/proxy/manifests.go:29]:
```suggestion
    app: kube-proxy # deprecated
    app.kubernetes.io/component: kube-proxy
```

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (sftim) [inline cmd/kubeadm/app/phases/addons/proxy/manifests.go:61]:
```suggestion
    k8s-app: kube-proxy # deprecated
    app.kubernetes.io/name: kube-proxy
```

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (sftim) [inline cmd/kubeadm/app/phases/addons/proxy/manifests.go:74]:
```suggestion
        k8s-app: kube-proxy # deprecated
        app.kubernetes.io/name: kube-proxy
```

**Review by sftim — CHANGES_REQUESTED** [review]
- We ~should~ might want to update `matchLabels` too. Are we happy to match on the preferred labels instead?
  - Ah, but this isn't an easy change to make. Might be nice to use the recommended labels just for **new** clusters.
- Let's set a label on the `coredns` ConfigMap
  - and the ClusterRole / ClusterRoleBinding
  - and the ServiceAccount

Optionally, set a `kubernetes.io/description` annotation on some of these objects too (I'd do the ones that have Pod templates) - will need a tweak to the PR title, or could happen in a separate PR.

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (paologallinaharbur) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go:61]: Deprecation node added

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (paologallinaharbur) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go:74]: same as above

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (paologallinaharbur) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go:29]: same as above

**Comment by paologallinaharbur** [comment]
Thanks for the review @sftim. 
I agree that it would be better to update `matchLabels` as well, but there is the risk of breaking the upgrade path. That is why I did not include in this PR 😕  

And I agree as well that `kubernetes.io/description` would be a nice addition, however, I would address it in a following PR to make sure the scope of this is as small as possible 😄

**Comment by paologallinaharbur** [comment]
I agree with you, I do not see much value added by the `# deprecated` note. The more I look at it, the more I think that it could be even confusing. It is not clear what are the implications to have that marked as deprecated and likely the text could be forgotten there for a lot of time. 😕 

If [sftim](https://github.com/sftim) agrees I would revert that and keep merely the `app.kubernetes.io/name` recommended label for now without modifying the matchLabels.

**Comment by paologallinaharbur** [comment]
Taking also into consideration this [doc page](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/) stating:
>Note: The CoreDNS Service is named kube-dns in the metadata.name field.
The intent is to ensure greater interoperability with workloads that relied on the legacy kube-dns Service name to resolve addresses internal to the cluster. Using a Service named kube-dns abstracts away the implementation detail of which DNS provider is running behind that common name.

 - I left just the `app.kubernetes.io/name` change avoiding updating the match labels to avoid any possible breaking change
 - As we agreed I reverted the `# deprecated` note since `kube-dns` name it is still the name of the service despite the `coreDNS` is used

@pacoxu do you think that we can move it forward?

**Comment by sftim** [comment]
> If [sftim](https://github.com/sftim) agrees I would revert that and keep merely the `app.kubernetes.io/name` recommended label for now without modifying the matchLabels.

OK, sure.

Maybe add a note next to the _official_ label. But that's optional.

**Comment by paologallinaharbur** [comment]
I understand that it is not yet approved even if @pacoxu marked it as LGTM. 

Do you have any suggestions to move this forward? Is there anything missing or someone I should ping?

**Review by neolit123 — COMMENTED** [review]
i am not convinced the app.k8s.io label is appropriate here, since we treat coredns as a system addon / component. not like an app.

also what about kube proxy?

@SataQiu WDYT?

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (sftim) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go:29]: We could register a brand new label to use instead. It depends what we want to declare, of course.

**Comment by sftim** [comment]
The thing _I'd_ like to see here is a shift away from using private labels in something that we expect people to install onto another cluster. See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set for an explanation of what “private” means here.

If we can't even begin that shift, is there any fix worth making?

**Comment by paologallinaharbur** [comment]
Thanks @neolit123 for the feedback 😄 

Quoting the [docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/):
> You can visualize and manage Kubernetes objects with more tools than kubectl. A common set of labels allows tools to work interoperably, describing objects in a common manner that all tools can understand.
In addition to supporting tooling, the recommended labels describe applications in a way that can be queried.
 [...]  applications are informal and described with metadata. The definition of what an application contains is loose.
 
IMO even if it is an addon/component it could be considered as well as an app that a user could be interested in monitoring and querying its metadata leveraging the recommended labels.
Prometheus annotation, for example, is enabled by default, and having such a standard label could help a lot when creating dashboards to discriminate scraped metrics having the same name, but coming from different services (`go_`, `proc_`,...).

And @sftim I would like to move away as well from private labels... I believe that the first step is adding the recommended ones

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (paologallinaharbur) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:29]: I would personally stick with the recommended label and not create brand-new ones. At least for our use case, it is important that all different workloads possess one standard label

**Pull request closed without merging**
